### PR TITLE
First set of ImageIO CGImageSource APIs

### PR DESCRIPTION
**Pull request by vkvenkat**
Adding initial CGImageSource APIs for the ImageIO framework. This includes APIs for extracting frames, creating thumbnails and associated unit tests for common image formats.

### CODE REVIEW

File: Frameworks/ImageIO/CGImageSource.mm [hunk 1]
```diff
@@ -25,50 +28,327 @@
...
+
+    return imageFormat;
+}
+@end
 
 /**
- @Status Stub
- @Notes
+ @Status Caveat
+ @Notes The current implementation supports common image file formats such as JPEG, GIF, TIFF, BMP, PNG and ICO.
```
Comment (jaredhms) [inline Frameworks/ImageIO/CGImageSource.mm]: nit: caveat should also include lack of support for kCGImageSourceTypeIdentifierHint

Comment (vkvenkat) [inline Frameworks/ImageIO/CGImageSource.mm]: Fixed

File: Frameworks/ImageIO/CGImageSource.mm [hunk 2]
```diff
@@ -25,50 +28,327 @@
...
+
+    RETURN_NULL_IF_FAILED(imageFormatConverter->Initialize(imageFrame.Get(), 
+                                                           GUID_WICPixelFormat32bppRGBA,
+                                                           WICBitmapDitherTypeNone, 
+                                                           nullptr, 
+                                                           0.f, 
+                                                           WICBitmapPaletteTypeCustom));
+
+    const unsigned int frameSize = frameWidth * frameHeight * 4;
+    std::vector<unsigned char> frameByteVector(frameSize);
```
Comment (jaredhms) [inline Frameworks/ImageIO/CGImageSource.mm]: std::vector doesn't let you transfer ownership of its buffer like this.  I think you'll need std::unique_ptr<unsigned char[]>, then call .release() on it after the successful copy into frameData.

File: Frameworks/ImageIO/CGImageSource.mm [hunk 3]
```diff
@@ -25,50 +28,327 @@
...
 */
 CGImageRef CGImageSourceCreateThumbnailAtIndex(CGImageSourceRef isrc, size_t index, CFDictionaryRef options) {
-    UNIMPLEMENTED();
-    return StubReturn();
+    RETURN_NULL_IF(!isrc);
+    NSData* imageData = ((ImageSource*)isrc).data;
+    RETURN_NULL_IF(!imageData);
+    RETURN_NULL_IF(index > (CGImageSourceGetCount(isrc) - 1));
+
+    MULTI_QI imageQueryInterface = {0};
```
Comment (jaredhms) [inline Frameworks/ImageIO/CGImageSource.mm]: Worth pulling out some of these identical blocks of code as helper functions?

File: include/ImageIO/CGImageSourceInternal.h [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+// THE SOFTWARE.
+//
+//******************************************************************************
+#pragma once
+
+#import <Foundation/Foundation.h>
+#import <CoreGraphics/CGDataProvider.h>
+#import <ImageIO/ImageIO.h>
+#import <ImageIO/CGImageSource.h>
+#import <objc/runtime.h>
```
Comment (jaredhms) [inline include/ImageIO/CGImageSourceInternal.h]: most (all?) of these can move to the .mm file.

**Comment by vkvenkat** [comment]
Thanks for the suggestions. We have incorporated the feedback and updated the PR.

**Pull request merged**